test/test_lazy.py

```diff
@@ -0,0 +1,202 @@
+#         return _gradient_operator(
+#             discr, DISCR_TAG_BASE, alpha, boundaries, u)
+
+#     compiled_op = actx.compile(op)
+#     alpha = discr.zeros(actx) + 1
+#     u = discr.zeros(actx)
+#     compiled_op(alpha, u)
+
+
+# FIXME: Re-enable and fix up if/when standalone divergence operator exists
+# def test_lazy_op_divergence(ctx_factory):
+#     cl_ctx = ctx_factory()
+#     actx, discr = _op_test_fixture(cl_ctx)
+#
+#     from grudge.dof_desc import DTAG_BOUNDARY, DISCR_TAG_BASE
+#     from mirgecom.diffusion import (
+#         _divergence_alpha_operator,
+#         DirichletDiffusionBoundary,
+#         NeumannDiffusionBoundary)
+#
+#     boundaries = {
+#         DTAG_BOUNDARY("x"): DirichletDiffusionBoundary(0.),
+#         DTAG_BOUNDARY("y"): NeumannDiffusionBoundary(0.)
+#     }
+#
+#     def op(alpha, u):
+#         return _divergence_alpha_operator(
+#             discr, DISCR_TAG_BASE, alpha, boundaries, u)
+
+#     compiled_op = actx.compile(op)
+#     alpha = discr.zeros(actx) + 1
+#     u = make_obj_array([discr.zeros(actx) for _ in range(2)])
+#     compiled_op(alpha, u)
+
+
+def test_lazy_op_diffusion(ctx_factory):
+    cl_ctx = ctx_factory()
+    actx, discr = _op_test_fixture(cl_ctx)
+    nodes = thaw(actx, discr.nodes())
+
+    from grudge.dof_desc import DTAG_BOUNDARY, DISCR_TAG_BASE
+    from mirgecom.diffusion import (
+        diffusion_operator,
+        DirichletDiffusionBoundary,
+        NeumannDiffusionBoundary)
+
+    boundaries = {
+        DTAG_BOUNDARY("x"): DirichletDiffusionBoundary(0.),
+        DTAG_BOUNDARY("y"): NeumannDiffusionBoundary(0.)
+    }
+
+    def op(alpha, u):
+        return diffusion_operator(
+            discr, DISCR_TAG_BASE, alpha, boundaries, u)
+
+    compiled_op = actx.compile(op)
+
+    alpha = discr.zeros(actx) + 1
+    u = actx.np.cos(np.pi*nodes[0])
+
+    rel_linf_error = partial(_rel_linf_error, discr)
+
+    eager_result = op(alpha, u)
+    lazy_result = compiled_op(alpha, u)
+    assert rel_linf_error(lazy_result, eager_result) < 1e-12
+
+
+def test_lazy_op_euler(ctx_factory):
+    cl_ctx = ctx_factory()
+    actx, discr = _op_test_fixture(cl_ctx)
+    nodes = thaw(actx, discr.nodes())
+
+    from grudge.dof_desc import DTAG_BOUNDARY
+    from mirgecom.eos import IdealSingleGas
+    from mirgecom.boundary import AdiabaticSlipBoundary
+    from mirgecom.euler import euler_operator
+
+    eos = IdealSingleGas()
+
+    boundaries = {
+        DTAG_BOUNDARY("x"): AdiabaticSlipBoundary(),
+        DTAG_BOUNDARY("y"): AdiabaticSlipBoundary()
+    }
+
+    def op(cv):
+        return euler_operator(discr, eos, boundaries, cv)
+
+    compiled_op = actx.compile(op)
+
+    from mirgecom.initializers import MulticomponentLump
+    init = MulticomponentLump(
+        dim=2, velocity=np.ones(2), spec_y0s=np.ones(3), spec_amplitudes=np.ones(3))
+    state = init(nodes)
+
+    rel_linf_error = partial(_rel_linf_error, discr)
+
+    eager_result = op(state)
```

Comment:
I think using the `PyOpenCLArrayContext` for the `eager_result` would probably be better. Right now, this seems to be computing the difference between an operator and its compiled counterpart. Although checking them is a good idea, but the non-compiled operator isn't really eagerly computed.

Reply:
How does it look now? (Is it allowed to share a `discr` between two array contexts? What about my lazy-to-eager result conversion, does that look ok?)